### DISCUSSION TITLE
How to use `locale-gen` with brew's dependency (glibc)?

### DISCUSSION THREAD

**Discussion by sisrfeng** [Homebrew/discussions#3682] [Linux] [answered]
My problem:
1.  How to get rid of the warning?
![image](https://user-images.githubusercontent.com/53520949/191932601-b5b26db7-24f4-4788-ab57-82f6d3bcf20d.png)

2. under nvim,  I hope that  `echo strftime('%H:%M')` will display the time in Chinese style, but it does not.

3. How to make `date` display the time in Chinese style?
![image](https://user-images.githubusercontent.com/53520949/191931394-190b6ba7-88f2-4574-903d-b0819d3070e9.png)

## Possible reason
There is `locale`  binary under `/home/linuxbrew`,  but no `locale-gen`.
(and `brew install locale-gen` fails), so `sudo locale-gen zh_CN.UTF-8`  is using `/usr/sbin/locale-gen`

![image](https://user-images.githubusercontent.com/53520949/191932040-db7f28ae-6a9d-4c38-85fc-d2b4e7a4c282.png)

![image](https://user-images.githubusercontent.com/53520949/191931472-90062af3-4f3b-4627-99f6-81e25454e51c.png)

**Comment by carlocab** [comment]
What is the output of `brew config` and `brew doctor`?

Please [don't use screenshots](https://meta.stackoverflow.com/a/285557) when sharing the output of commands in your terminal. They're often difficult to read and impossible to reuse (e.g. when looking up error messages). Instead, copy the text here and wrap them in three back-ticks (```) on the line before and on the line after to format them properly.

CC @Homebrew/linux

**Comment by sisrfeng** [comment]
brew config
```
HOMEBREW_VERSION: 3.6.1
ORIGIN: https://mirrors.tuna.tsinghua.edu.cn/git/homebrew/brew.git
HEAD: 6e2b162c4786e075323f038d46bfb566d91889e7
Last commit: 13 days ago
Core tap ORIGIN: https://mirrors.tuna.tsinghua.edu.cn/git/homebrew/homebrew-core.git
Core tap HEAD: 4f38944f7f5afeda86609f8bcb7b944400fc6025
Core tap last commit: 9 days ago
Core tap branch: master
HOMEBREW_PREFIX: /home/linuxbrew/.linuxbrew
HOMEBREW_BAT: set
HOMEBREW_BOTTLE_DOMAIN: https://mirrors.tuna.tsinghua.edu.cn/homebrew-bottles
HOMEBREW_BREW_GIT_REMOTE: https://mirrors.tuna.tsinghua.edu.cn/git/homebrew/brew.git
HOMEBREW_BROWSER: w3m
HOMEBREW_CASK_OPTS: []
HOMEBREW_CORE_GIT_REMOTE: https://mirrors.tuna.tsinghua.edu.cn/git/homebrew/homebrew-core.git
HOMEBREW_DISPLAY: 192.168.100.94:0
HOMEBREW_EDITOR: nvim
HOMEBREW_MAKE_JOBS: 24
HOMEBREW_NO_AUTO_UPDATE: set
HOMEBREW_NO_ENV_HINTS: set
Homebrew Ruby: 2.6.8 => /home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/vendor/portable-ruby/2.6.8_1/bin/ruby
CPU: 24-core 64-bit skylake
Clang: 14.0.6
Git: 2.37.3 => /home/linuxbrew/.linuxbrew/bin/git
Curl: 7.68.0 => /usr/bin/curl
Kernel: Linux 5.4.0-125-generic x86_64 GNU/Linux
OS: Ubuntu 20.04.5 LTS (focal)
Host glibc: 2.31
/usr/bin/gcc: 9.4.0
/usr/bin/ruby: 2.7.0
glibc: 2.35
gcc@11: 11.3.0
xorg: N/A
```

**Comment by sisrfeng** [comment]
brew doctor
```
Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry or file an issue; just ignore this. Thanks!

Warning: Some installed formulae are deprecated or disabled.
You should find replacements for the following formulae:
  googler

Warning: "config" scripts exist outside your system or Homebrew directories.
`./configure` scripts often look for *-config scripts to determine if
software packages are installed, and which additional flags to use when
compiling and linking.

Having additional scripts in your path can confuse software installed via
Homebrew if the config script overrides a system or Homebrew-provided
script of the same name. We found the following "config" scripts:
  /home/wf/d/anaconda3/envs/t_pool3/bin/python3-config
  /home/wf/d/anaconda3/envs/t_pool3/bin/python3.9-config
  /home/wf/d/anaconda3/envs/t_pool3/bin/libpng-config
  /home/wf/d/anaconda3/envs/t_pool3/bin/libpng16-config
  /home/wf/d/anaconda3/envs/t_pool3/bin/ncursesw6-config
  /home/wf/d/anaconda3/envs/t_pool3/bin/freetype-config
  /home/wf/d/anaconda3/bin/curl-config
  /home/wf/d/anaconda3/bin/python3-config
  /home/wf/d/anaconda3/bin/python3.9-config
  /home/wf/d/anaconda3/bin/onig-config
  /home/wf/d/anaconda3/bin/libpng-config
  /home/wf/d/anaconda3/bin/xml2-config
  /home/wf/d/anaconda3/bin/libpng16-config
  /home/wf/d/anaconda3/bin/krb5-config
  /home/wf/d/anaconda3/bin/icu-config
  /home/wf/d/anaconda3/bin/ncursesw6-config
  /home/wf/d/anaconda3/bin/pcre-config
  /home/wf/d/anaconda3/bin/xslt-config
  /home/wf/d/anaconda3/bin/freetype-config

Warning: Unbrewed header files were found in /usr/local/include.
If you didn't put them there on purpose they could cause problems when
building Homebrew formulae, and may need to be deleted.

Unexpected header files:
  /usr/local/include/Eigen/src/Cholesky/LDLT.h
  /usr/local/include/Eigen/src/Cholesky/LLT.h
  /usr/local/include/Eigen/src/Cholesky/LLT_LAPACKE.h
  /usr/local/include/Eigen/src/CholmodSupport/CholmodSupport.h
  /usr/local/include/Eigen/src/Core/Array.h
  /usr/local/include/Eigen/src/Core/ArrayBase.h
  /usr/local/include/Eigen/src/Core/ArrayWrapper.h
  /usr/local/include/Eigen/src/Core/Assign.h
  /usr/local/include/Eigen/src/Core/AssignEvaluator.h
  /usr/local/include/Eigen/src/Core/Assign_MKL.h
  /usr/local/include/Eigen/src/Core/BandMatrix.h
  /usr/local/include/Eigen/src/Core/Block.h
  /usr/local/include/Eigen/src/Core/BooleanRedux.h
  /usr/local/include/Eigen/src/Core/CommaInitializer.h
  /usr/local/include/Eigen/src/Core/ConditionEstimator.h
  /usr/local/include/Eigen/src/Core/CoreEvaluators.h
  /usr/local/include/Eigen/src/Core/CoreIterators.h
  /usr/local/include/Eigen/src/Core/CwiseBinaryOp.h
  /usr/local/include/Eigen/src/Core/CwiseNullaryOp.h
  /usr/local/include/Eigen/src/Core/CwiseTernaryOp.h
  /usr/local/include/Eigen/src/Core/CwiseUnaryOp.h
  /usr/local/include/Eigen/src/Core/CwiseUnaryView.h
  /usr/local/include/Eigen/src/Core/DenseBase.h
  /usr/local/include/Eigen/src/Core/DenseCoeffsBase.h
  /usr/local/include/Eigen/src/Core/DenseStorage.h
  /usr/local/include/Eigen/src/Core/Diagonal.h
  /usr/local/include/Eigen/src/Core/DiagonalMatrix.h
  /usr/local/include/Eigen/src/Core/DiagonalProduct.h
  /usr/local/include/Eigen/src/Core/Dot.h
  /usr/local/include/Eigen/src/Core/EigenBase.h
  /usr/local/include/Eigen/src/Core/ForceAlignedAccess.h
  /usr/local/include/Eigen/src/Core/Fuzzy.h
  /usr/local/include/Eigen/src/Core/GeneralProduct.h
  /usr/local/include/Eigen/src/Core/GenericPacketMath.h
  /usr/local/include/Eigen/src/Core/GlobalFunctions.h
  /usr/local/include/Eigen/src/Core/IO.h
  /usr/local/include/Eigen/src/Core/Inverse.h
  /usr/local/include/Eigen/src/Core/Map.h
  /usr/local/include/Eigen/src/Core/MapBase.h
  /usr/local/include/Eigen/src/Core/MathFunctions.h
  /usr/local/include/Eigen/src/Core/MathFunctionsImpl.h
.....
(lots of similar lines)
.....
  /usr/local/include/eigen3/unsupported/Eigen/src/SpecialFunctions/SpecialFunctionsPacketMath.h
  /usr/local/include/eigen3/unsupported/Eigen/src/SpecialFunctions/arch/CUDA/CudaSpecialFunctions.h
  /usr/local/include/eigen3/unsupported/Eigen/src/Splines/Spline.h
  /usr/local/include/eigen3/unsupported/Eigen/src/Splines/SplineFitting.h
  /usr/local/include/eigen3/unsupported/Eigen/src/Splines/SplineFwd.h

Warning: Unbrewed static libraries were found in /usr/local/lib.
If you didn't put them there on purpose they could cause problems when
building Homebrew formulae, and may need to be deleted.

Unexpected static libraries:
  /usr/local/lib/libceres.a
  /usr/local/lib/libceres_cuda_kernels.a
```
end of the output

**Comment by sisrfeng** [comment]
Hi, any update?
Thanks!

**Reply by carlocab** [reply]
Ping @Homebrew/linux

**Reply by sisrfeng** [reply]
This page is not found :  https://github.com/orgs/Homebrew/teams/linux

**Comment by danielnachun** [comment]
Maybe this is just because `glibc` is keg-only?  What happens if you try `brew link --force glibc`?

**Reply by sisrfeng** [reply]
After these commands, the problem 'perl: warning: Falling back to the standard locale ("C")' is still there:

```sh
$ brew link --force glibc
Warning: Already linked: /home/linuxbrew/.linuxbrew/Cellar/glibc/2.35
To relink, run:
  brew unlink glibc && brew link glibc


$  brew unlink glibc && brew link glibc
Unlinking /home/linuxbrew/.linuxbrew/Cellar/glibc/2.35... 292 symlinks removed.
Linking /home/linuxbrew/.linuxbrew/Cellar/glibc/2.35... 294 symlinks created. 
```

**Reply by sisrfeng** [reply]
why `gcc` of brew can' t be found by `where`?
![image](https://user-images.githubusercontent.com/53520949/193740106-648048b6-a37b-44d1-8703-cdfb947fae5e.png)

I'm confused with brew's glibc and the system's glibc  and I find this: https://github.com/Homebrew/glibc-bootstrap/#motivation.
Any doc or FAQ  for newby user on brew's glibc?

PS: from `brew config`:
> Host glibc: 2.31
> /usr/bin/gcc: 9.4.0
> /usr/bin/ruby: 2.7.0
> glibc: 2.35

**Reply by carlocab** [reply]
Try `where gcc-12`.

**Reply by sisrfeng** [reply]
get:   /home/linuxbrew/.linuxbrew/bin/gcc-12

**Reply by cho-m** [reply]
I think this is the `glibc` postinstall issue we triaged in https://github.com/Homebrew/homebrew-core/issues/110812.

Essentially, we currently only set up US locale even if users added `LANG `/`LC_*` to environment.

If we fix postinstall and user has those variables when `glibc` is installed, then the locale data should be added automatically. And `brew postinstall glibc` could be used to refresh data

**Reply by sisrfeng** [reply]
Thx, looking forward to update in that issue

**Reply by sisrfeng** [reply]
Is the following  problem also caused by "Essentially, we currently only set up US locale even if users added LANG /LC_* to environment."?
before setting `export TZ=Asia/Shanghai`,  strftime() in nvim and `set -g status-right "#[fg=red,dim]   %-d日 - %H:%M "` in tmux shows time 8 hours left behind Shanghai's time, even though  I have `export LC_TIME=zh_CN.utf_8` in my .zshrc.

Is the command `brew postinstall glibc` a temporary  solution to similar problems?

**Reply by danielnachun** [reply]
I think the root cause of the problem right now is that `brew` filters out environment variables even if you've exported them.  So even when the user sets `LANG` or `TZ`, when `brew postinstall glibc` is actually run, those variables are not visible to `glibc` because they were filtered out.  The solution here is to add those variables to the allowlist in `brew`.

Currently it seems the most important ones are `LANG` and `TZ`.  Are there any environment variables other than those two that we think would need to be passed through as well?  I am in the US and only use US timezones and the US english locale so I am not sure which other variables are needed beyond those two.

**Reply by sisrfeng** [reply]
What's allowlist ? Can not be found in the doc and man.

**Reply by danielnachun** [reply]
It's not documented to my knowledge, but the list is here: https://github.com/Homebrew/brew/blob/fb52a12144e80118c112ea744ab65ab6ca997d67/bin/brew#L130-L135.  Any environment variable not in the list is ignored when `brew` is run.  Feel free to open a PR to add `LANG` and `TZ` to that list.  We don't want to go too crazy with adding too many variables to that list, but those two seem reasonable as they are something we expect the user may wish to set.

**Comment by lestephane** [comment, marked as answer]
I was able to successfully add a locale I needed (`de_DE.UTF-8` in my case) by doing

```
$ localedef -v -i de_DE -f UTF-8 de_DE.UTF-8
$ locale -a
C
de_DE.utf8
en_US.utf8
POSIX
```
both commands are from homebrew for linux
```
$ which localedef locale
/home/linuxbrew/.linuxbrew/bin/localedef
/home/linuxbrew/.linuxbrew/bin/locale
```

You'll probably need to run the same `localedef` adjusted for `zh_CN.UTF-8`

**Reply by sisrfeng** [reply]
Done!

Summary 
1. locale-gen is for the Debian GNU/Linux system (but may be used by others).  
`brew install locale-gen` fails.
![image](https://user-images.githubusercontent.com/53520949/193961208-8ab8c351-5e72-4402-be05-30066c0e4dd9.png)



2. `man localedef` shows that "This page is part of release 5.05 of the Linux man-pages project"
![image](https://user-images.githubusercontent.com/53520949/193961117-8ca73007-d7b5-40ba-a300-f60b6b902a28.png)

**Reply by sisrfeng** [reply]
Debian's locale-gen  is a thin wrapper on   GNU/Linux' s  localedef, but is not necessary, right?

**Reply by sisrfeng** [reply]
Further reading:

www.kernel.org/doc/man-pages   and GNU

https://superuser.com/a/879978/1651425
> For example, the manual pages for socket, send, and recv (in general anything in section 2) are part of the kernel and documented as part of the kernel documentation. While gethostbyname is a glibc function, it too is documented as part of the kernel.org manpages on my system. I suspect this is because the GNU people are not interested in manpages as a primary documentation format, and focus their efforts on Texinfo.

**Reply by sisrfeng** [reply]
Some useful information printed by `brew install glibc`  (after my `brew remove glibc`)

==> Pouring glibc-2.35_1.x86_64_linux.bottle.tar.gz
==> /home/linuxbrew/.linuxbrew/Cellar/glibc/2.35_1/sbin/ldconfig
==> Installing locale data for en_US.UTF-8
==> /home/linuxbrew/.linuxbrew/Cellar/glibc/2.35_1/bin/localedef -i en_US -f UTF-8 en_US.UTF-8
==> Caveats
The Homebrew's Glibc has been installed with the following executables:
  /home/linuxbrew/.linuxbrew/opt/glibc/bin/ldd
  /home/linuxbrew/.linuxbrew/opt/glibc/bin/ld.so
  /home/linuxbrew/.linuxbrew/opt/glibc/sbin/ldconfig

By default, Homebrew's linker will not search for the system's libraries. If you
want Homebrew to do so, run:

  cp "/home/linuxbrew/.linuxbrew/etc/ld.so.conf.d/99-system-ld.so.conf.example" "/home/linuxbrew/.linuxbrew/etc/ld.so.conf.d/99-system-ld.so.conf"
  brew postinstall glibc

to append the system libraries to Homebrew's ld search paths. This is risky and
**highly not recommended**, because it may cause linkage to Homebrew libraries
mixed with system libraries.

glibc is keg-only, which means it was not symlinked into /home/linuxbrew/.linuxbrew,
because it can shadow system glibc if linked.

If you need to have glibc first in your PATH, run:
  echo 'export PATH="/home/linuxbrew/.linuxbrew/opt/glibc/bin:$PATH"' >> ~/.zshrc
  echo 'export PATH="/home/linuxbrew/.linuxbrew/opt/glibc/sbin:$PATH"' >> ~/.zshrc

For compilers to find glibc you may need to set:
  export LDFLAGS="-L/home/linuxbrew/.linuxbrew/opt/glibc/lib"
  export CPPFLAGS="-I/home/linuxbrew/.linuxbrew/opt/glibc/include"

**Reply by sisrfeng** [reply]
Note: there is no `/home/linuxbrew/.linuxbrew/bin/localedef` in 2 of my 3 linux machines. 
In all of the machines, when I cd to  `/home/linuxbrew/.linuxbrew/opt`, and after `ll`, there is `glibc -> ../Cellar/glibc/2.35_1`.

**Reply by sisrfeng** [reply]
`localedef` is under `Cellar/glibc/2.35_1/bin/`
